### ISSUE TITLE
Improvements to jiratoken

### ISSUE THREAD

**Issue by CameronLonsdale** [trufflesecurity/trufflehog#649]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Firstly, it's helpful to know that the "jiratoken" or Atlassian API token will also work on Confluence Cloud.

Secondly, there are a few improvements that can be made to your detector in https://github.com/trufflesecurity/trufflehog/blob/198cb1a786a97730e9532260f80fd1f2d795bfa2/pkg/detectors/jiratoken/jiratoken.go

## Description of the Preferred Solution

For this regex https://github.com/trufflesecurity/trufflehog/blob/198cb1a786a97730e9532260f80fd1f2d795bfa2/pkg/detectors/jiratoken/jiratoken.go#L25, you detect for 24 characters which is correct. However the structure of the token is that the last four characters is a snippet of the hash of the first 20 characters. That can be a fast way to rule out false positives. Take the first 20 characters of the token, hash it using MD5 and then take the first 4 hex characters of the hash and compare it to validate the token.

Secondly, using the dashboard rest API is not a great way to validate a token because i think it will always work if the jira dashboard is open to anonymous users. https://github.com/trufflesecurity/trufflehog/blob/198cb1a786a97730e9532260f80fd1f2d795bfa2/pkg/detectors/jiratoken/jiratoken.go#L69. What would be more robust instead is a request to the `/gateway/api/graphql` endpoint, which can be done with `https://+resDomain+/gateway/api/graphql` or simply `https://atlassian.net/gateway/api/graphql` will work too. You could use a query like `query verify { me { user {name}}}` to get the name of the account holder for that token. Or you could also do `query verify { me { user { name, ... on AtlassianAccountUser {email}} }}` to get the email address of the account holder as well for output.

### References

The graphql documentation is available in the docs tab of https://atlassian.net/gateway/api/graphql.

**Comment by dustin-decker** [comment]
Great information @CameronLonsdale. My apologies that this has sat so long. We'll look at making these improvements!

**Comment by jsolis** [comment]
https://github.com/trufflesecurity/trufflehog/pull/1288

This PR fixes this issue FYI

**Comment by dustin-decker** [comment]
There's still a few improvements from this issue that'd we'd like to implement.

**Comment by nyanshak** [comment]
With the 24-character tokens (old format):

```go
			prefix := resToken[0:20]
			suffix := resToken[20:24]

			// Structure of the token is that the last four characters are a
			// the first four characters of the md5sum of the first 20
			// characters.
			//
			// If the hash is not valid, skip token as it is a false positive.
			hasValidHash := string(suffix) == strings.ToUpper(fmt.Sprintf("%x", md5.Sum([]byte(prefix))))[0:4]
			if !hasValidHash {
				continue
			}
```

To use graphql endpoint:

```go
// Change to use /gateway/api/graphql
// Also always try atlassian.net, in addition to any found URLs

const AtlassianAPIValidationQuery = `{"query":"query validation {me{user{accountId,... on AtlassianAccountUser {email}}}}", "variables": null, "operationName": "validation" }`

type AtlassianAPIGraphqlMeResponse struct {
	Data struct {
		Me struct {
			User struct {
				AccountID string `json:"accountId"`
				Email     string `json:"email"`
			} `json:"user"`
		} `json:"me"`
	} `json:"data"`
}

// Parse the json returned into ^ this struct, then you can add `AccountID` to `ExtraData` :)

// Fail if status != 200 or != 401
```

---

I was going to start contributing the above, but then I realized that the tokens being generated (https://id.atlassian.com/manage-profile/security/api-tokens) are a different format (longer than 24 characters and different method of calculating the hashes). I don't know if/where the format is publicly documented, otherwise I would try to contribute this as well. Note that the new format is not just a longer version of the previous one, as it looks to be some prefix + base64-url-encoded-string + checksum.

![image](https://user-images.githubusercontent.com/3431490/235788129-a8732406-a7ae-40c7-8e44-7ae8e1aa9c23.png)

The above changes would still apply to old tokens, so may still be worth working through.

---

On a side note, @CameronLonsdale mentions:

> Firstly, it's helpful to know that the "jiratoken" or Atlassian API token will also work on Confluence Cloud.

It may be worth renaming this detector to something like `AtlassianAPI`, and changing `Keywords` and the list of strings in the `detectors.PrefixRegex` calls to be something like: `[]string{"atlassian", "confluence", "jira"}`.

**Comment by CameronLonsdale** [comment]
@nyanshak Yeah it's a prefix, data, then a checksum for the new format. Its likely we no longer need prefix regex if we can correctly identify tokens with (a) checksum in the old pattern or (b) token prefix + checksum in the new pattern.

I managed to reverse engineer in cyberchef:

`ATATT3xFfGF.......vpfo=9D1CE27B` is an example. The check sum is the final 8 characters using a CRC-32 algorithm from everything including the start of the prefix up until the start of the `9D1CE27B`

I'm not sure where the prefix ends, I think it might be `ATATT` ?

**Comment by CameronLonsdale** [comment]
Bump on fixing this one, still getting false positives because of the detection just looking at the default dashboard

**Comment by kashifkhan0771** [comment]
Hi @CameronLonsdale, thank you for your patience and for following up on this. I'll look into it as soon as I can.

**Comment by kashifkhan0771** [comment]
I’ve opened a [pull request](https://github.com/trufflesecurity/trufflehog/pull/4155) that updates the verification API and the keywords list.

> However the structure of the token is that the last four characters is a snippet of the hash of the first 20 characters. That can be a fast way to rule out false positives. Take the first 20 characters of the token, hash it using MD5 and then take the first 4 hex characters of the hash and compare it to validate the token.

Regarding your suggestion to verify the token validity using a hashing method. I wasn’t able to find any official documentation confirming that approach. While it might work, I’m hesitant to implement it without proper validation, especially since JIRA is a widely used detector. Introducing an unverified method could lead to unintended issues. If you have any sources confirming that please share.

**Comment by CameronLonsdale** [comment]
That's alright, It may not be documented 🤔 , but you can reverse engineer it. Your PR is already a big improvement. Thank you!
